### PR TITLE
[mlir][vector][spirv] Lower vector.load and vector.store to SPIR-V

### PR DESCRIPTION
**Pull request by Hsiangkai**
Add patterns to lower vector.load to spirv.load and vector.store to spirv.store.

### CODE REVIEW

File: mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp [hunk 1]
```diff
@@ -509,6 +509,72 @@ struct VectorShuffleOpConvert final
...
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(vector::LoadOp loadOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto memrefType = loadOp.getMemRefType();
+    auto attr =
+        dyn_cast_or_null<spirv::StorageClassAttr>(memrefType.getMemorySpace());
+    if (!attr)
+      return failure();
```
Comment (antiagainst) [inline mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp:523]: It would be nice to add useful messages why matching is failure, like `return rewriter.notifyMatchFailure(loadOp, "expected spirv.storage_class memory space")`. Right now that's missing in other patterns due to historical reasons--the mechanims for doing this was added later after developing the patterns.

File: mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp [hunk 2]
```diff
@@ -509,6 +509,72 @@ struct VectorShuffleOpConvert final
...
+
+    const auto &typeConverter = *getTypeConverter<SPIRVTypeConverter>();
+    auto loc = loadOp.getLoc();
+    Value accessChain =
+        spirv::getElementPtr(typeConverter, memrefType, adaptor.getBase(),
+                             adaptor.getIndices(), loc, rewriter);
+    if (!accessChain)
+      return failure();
```
Comment (antiagainst) [inline mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp:531]: Similarly, `return rewriter.notifyMatchFailure(loadOp, "failed to get memref element pointer")`.

File: mlir/test/Conversion/VectorToSPIRV/vector-to-spirv.mlir [hunk 3]
```diff
@@ -631,3 +631,105 @@ func.func @shape_cast_size1_vector(%arg0 : vector<f32>) -> vector<1xf32> {
   %1 = vector.shape_cast %arg0 : vector<f32> to vector<1xf32>
   return %1 : vector<1xf32>
 }
+
+// -----
+
+module attributes {
+  spirv.target_env = #spirv.target_env<
+    #spirv.vce<v1.0, [Shader], [SPV_KHR_storage_buffer_storage_class]>, #spirv.resource_limits<>>
```
Comment (antiagainst) [inline mlir/test/Conversion/VectorToSPIRV/vector-to-spirv.mlir:639]: Hmm, this kinda of pointer bitcasting typically is not valid in `Shader` capabilities--it would nomrally require `Kernel` capabilities. You can find details in https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_validation_rules (warning: very hard to parse through..) Are you actually using these patterns for graphics APIs like Vulkan? Or are you doing it for compute APIs like OpenCL?

We probably need to add a validation rule in `spirv.module` op for this; but here we aren't using `spirv.module` op anyway so it won't kick in. I'd suggest to make it `#spirv.vce<v1.0, [Kernel], []>` to be clear though.

**Review by antiagainst — CHANGES_REQUESTED** [review]
Looks nice! Just a few nits. Thanks!

File: mlir/test/Conversion/VectorToSPIRV/vector-to-spirv.mlir (hunk 3 above)
Comment (Hsiangkai) [reply mlir/test/Conversion/VectorToSPIRV/vector-to-spirv.mlir:639]: Yes, I intend to use these patterns for graphics APIs. I tried to change it to `[Kernel]`. However, I got the legalization issue after conversion.
```
spirv.AccessChain illegal: requires at least one capability in [Shader] but none allowed in target environment
```

Comment (antiagainst) [reply mlir/test/Conversion/VectorToSPIRV/vector-to-spirv.mlir:639]: Okay. On Vulkan it'd need extra capabilities like physical storage buffer.. Anyway, that is beyond what's being done in this patch; so I'm not gonna hold this patch onto that. :)

**Review by antiagainst — APPROVED** [review]
